text/0517-io-os-reform.md

```diff
@@ -473,12 +635,289 @@ throughout IO, we can go on to explore the modules in detail.
+
+    // Adapt `self` to yield only the first `limit` bytes
+    fn take(self, limit: u64) -> Take<Self> { ... }
+
+    // Whenever reading from `self`, push the bytes read to `out`
+    fn tee<W: Writer>(self, out: W) -> Tee<Self, W> { ... }
+}
+
+trait WriteExt: Write {
+    // ... eliding the methods already described above
+
+    // Reify a borrowed writer as owned
+    fn by_ref<'a>(&'a mut self) -> ByRef<'a, Self> { ... }
+
+    // Whenever bytes are written to `self`, write them to `other` as well
+    fn broadcast<W: Writer>(self, other: W) -> Broadcast<Self, W> { ... }
+}
+
+// An adaptor converting an `Iterator<u8>` to `Read`.
+pub struct IterReader<T> { ... }
+```
+
+As with `std::iter`, these adapters are object unsafe and hence placed
+in an extension trait with a blanket `impl`.
+
+Note that the same `ByRef` type is used for both `Read` and `Write`
+-- and this RFC proposes to use it for `std::iter` as well. The
+insight is that there is no difference between the *type* used for
+by-ref adapters in any of these cases; what changes is just which
+trait defers through it. So, we propose to add the following to `core::borrow`:
+
+```rust
+pub struct ByRef<'a, Sized? T:'a> {
+    pub inner: &'a mut T
+}
+```
+
+which will allow `impl`s like the following in `core::io`:
+
+```rust
+impl<'a, W: Write> Write for ByRef<'a, W> {
+    #[inline]
+    fn write(&mut self, buf: &[u8]) -> Result<uint, W::Err> { self.inner.write(buf) }
+
+    #[inline]
+    fn flush(&mut self) -> Result<(), W::Err> { self.inner.flush() }
+}
+```
+
+#### Free functions
+[Free functions]: #free-functions
+
+The current `std::io::util` module also includes a number of primitive
+readers and writers, as well as `copy`. These are updated as follows:
+
+```rust
+// A reader that yields no bytes
+fn empty() -> Empty; // in theory just returns `impl Reader`
+
+// A reader that yields `byte` repeatedly (generalizes today's ZeroReader)
+fn repeat(byte: u8) -> Repeat;
+
+// A writer that ignores the bytes written to it (/dev/null)
+fn sink() -> Sink;
+
+// Copies all data from a Reader to a Writer
+pub fn copy<E, R, W>(r: &mut R, w: &mut W) -> Result<(), E> where
+    R: Read<Err = E>,
+    W: Write<Err = E>
+```
+
+#### Seeking
+[Seeking]: #seeking
+
+The seeking infrastructure is largely the same as today's, except that
+`tell` is renamed to follow the RFC's design principles and the `seek`
+signature is refactored with more precise types:
+
+```rust
+pub trait Seek {
+    type Err;
+    fn position(&self) -> Result<u64, Err>;
+
+    // returns the new position after seeking
+    fn seek(&mut self, pos: SeekPos) -> Result<u64, Err>;
+}
+
+pub enum SeekPos {
+    FromStart(u64),
+    FromEnd(i64),
+    FromCur(i64),
+}
+```
+
+#### Buffering
+[Buffering]: #buffering
+
+The current `Buffer` trait will be renamed to `BufferedRead` for
+clarity (and to open the door to `BufferedWrite` at some later
+point):
+
+```rust
+pub trait BufferedRead: Read {
```

Comment:
This trait name doesn't sit super well with me. With `Read` and `Write` we've got a clear convention for naming the trait (one method == one possible name). In this case the name of the trait, `BufferedRead` has no connection with its `fill_buf` method, and both names are a little odd to me.

I'd love to trim this down to one method if possible (not sure that it is) or perhaps one "main method" which the trait is named after. One possibility would be `trait ReadBuffer` and `fn read_buffer()` but those also don't sit super well with me.